### ISSUE TITLE
[RFC]: Reduced sampling for tensor-parallel decoding

### ISSUE THREAD

**Issue by hzx55906** [vllm-project/vllm#54333]
### Motivation.

Tensor-parallel vocabulary projection produces one vocabulary shard per TP rank. The current sampling path gathers all shards before selecting the next token, resulting in `O(B × V)` communication per sampling step.

This is unnecessary for several common workloads:

- Greedy decoding only needs the global argmax.
- Bounded top-k sampling only needs the global top-k candidate set.
- Greedy speculative verification only needs one target argmax per position.

This RFC proposes a generic reduced sampling path that avoids full-vocabulary communication when possible and safely falls back otherwise.

### Proposed Change.

Add an opt-in flag:

```text
--enable-reduced-sampling
```

The feature is disabled by default and initially requires:

- Model Runner V2.
- Tensor parallel size greater than one.
- Standard vocab-parallel logits heads.
- The built-in sampler.

### Local-logits capture

While `model.compute_logits()` runs, the common `LogitsProcessor` may skip the TP gather and expose:

- Shard-local logits.
- Global vocabulary size.
- The shard's starting global token ID.
- Local padded vocabulary size.

Local logits are accepted only when exactly one standard projection is captured and the returned Tensor matches the captured result.

Models with missing metadata, added vocabulary, multiple projections, or unsupported postprocessing use the normal full-logits path. This avoids adding `compute_logits_local()` to individual model files.

### Greedy sampling

Each rank computes its local maximum and global token ID. The ranks gather only these candidate pairs and select the global maximum.

```text
Full logits: O(B × V)
Reduced:     O(B × TP)
```

### Bounded random sampling

For finite top-k sampling, each rank gathers its local top-k values and global token IDs. Global top-k/top-p filtering and Gumbel sampling then run on the combined candidates.

```text
Full logits: O(B × V)
Reduced:     O(B × K × TP)
```

The reduced path is used only when its communication payload is smaller than gathering the local full-vocabulary logits.

Random requests with explicit seeds fall back because the reduced candidate positions do not preserve the existing full-vocabulary Gumbel RNG mapping.

### Speculative decoding

Reduced speculative verification is supported when:

- All target requests are greedy.
- Draft sampling is greedy.
- Synthetic conditional rates are disabled.
- The built-in rejection sampler is used.

A Triton kernel compares draft tokens with the reduced target argmaxes and returns the accepted prefix, first rejected target token, or bonus token.

Probabilistic target or draft sampling continues to use the existing full-vocabulary rejection sampler.

### Fallback policy

Fallback is evaluated for the whole sampler batch. Full logits are used when any active request needs:

- Logit bias, allowed-token filtering, bad words, or penalties.
- Thinking budgets, `min_p`, grammar, or structured output.
- Logprobs, explicit logprob token IDs, NaN counting, or distribution replay.
- Random sampling without bounded top-k or with an explicit seed.
- Probabilistic speculative decoding.
- Unsupported model/head metadata.
- Candidate communication that would not be smaller.

The configuration is rejected for Model Runner V1, TP=1, batch-sharded sampling, LoRA, diffusion models, and custom samplers.

### Correctness and limitations

For unique maxima, reduced greedy sampling returns the same global argmax as the full-logits path.

For bounded random sampling, the global top-k candidate set is preserved unless exact logits ties occur at the cutoff. The current full path may retain all cutoff ties, while reduced sampling selects exactly K local candidates. Handling these ties through detection or fallback remains an open question.

Random reduced sampling targets the same distribution outside tie cases but does not guarantee bit-for-bit equality with every existing backend.

The capture mechanism detects out-of-place model postprocessing. An out-of-tree model that mutates the same Tensor in place may require a future explicit local-logits API.

### Alternatives considered

- Model-specific `compute_logits_local()`: rejected as the primary design because it requires changes across model families.
- Batch-sharded sampling: retained as a separate, mutually exclusive strategy.
- Always gathering top-k candidates: rejected because large K can communicate more data than full logits.
- Probabilistic speculative reduction: deferred because exact rejection sampling requires complete probability distributions.

### Rollout and validation

1. Introduce the feature behind an opt-in flag.
2. Validate greedy and speculative exact-match behavior on TP=2+.
3. Validate random sampling distributions and every fallback condition.
4. Measure latency, throughput, NCCL traffic, and peak memory.
5. Add observability for reduced-path hits and fallback reasons.
6. Consider broader support or default enablement only after validation.

### Feedback Period.

Proposed period: August 29, 2026 through September 12, 2026.

### CC List.

TODO: Add Model Runner V2, sampler, speculative decoding, and distributed communication maintainers.

### Any Other Things.

Prototype PR: https://github.com/vllm-project/vllm/pull/54332

The implementation does not modify individual model files and retains full logits as the correctness and peak-memory fallback.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
